internal/generationtest/generation_test.bzl

```diff
@@ -0,0 +1,41 @@
+"""
+Test for generating rules from gazelle. 
+"""
+
+load("@io_bazel_rules_go//go:def.bzl", "go_test")
+
+def gazelle_generation_test(name, gazelle_binary_name, test_data, test_data_dir, gazelle_binary_dir = ""):
+    """
+    gazelle_generation_test is a macro for testing gazelle against workspaces.
+
+    Args:
+        name: the name of the test.
+        gazelle_binary_name: the name of the gazelle binary target. For example, if you have a
+        gazelle_binary target where you pass the name "my_gazelle", this would be "my_gazelle".
+        test_data: a glob of the test data files you will pass to the test.
+        test_data_dir: the workspace relative path where the testdata file is.
+        gazelle_binary_dir: the workspace relative path where the gazelle_binary target is. For example,
+        if "my_gazelle" is defined in a BUILD.bazel file in <wkspace>/my/path, this arg should be
+        "my/path".
+    """
+    native.genrule(
```

Comment:
This looks like a lot of work to figure out where the template is and all that stuff. What would you think about modifying `go_test` below to have some [`args`](https://docs.bazel.build/versions/main/be/common-definitions.html#common-attributes-tests)? We use stuff like this throughout the repo. For example [here](https://github.com/bazelbuild/rules_go/blob/bd0dd3fe48a53227f2c457e8c2097e1c097d1f38/tests/core/go_test/BUILD.bazel#L51-L61).

1. Change the `gazelle_generation_test` to be:

```
def gazelle_generation_test(name, gazelle_binary, test_data):
    """
    gazelle_generation_test is a macro for testing gazelle against workspaces.
    Args:
        name: the name of the test.
        gazelle_binary: A path to the binary under test. For example: "//path/to:my_gazelle".
        test_data: a target of the test data files you will pass to the test. This can be a
          https://docs.bazel.build/versions/main/be/general.html#filegroup
    """
   ```

2.  Remove the genrule and change the `go_test` to directly manage those inputs:

   ```
go_test(
        name = name,
        args = [
                "--gazelle_binary_path=$(location %s)" % gazelle_binary,
                "--test_data=$(locations %s)" % test_data,
        ],
        deps = [
            "//testtools",
            "@io_bazel_rules_go//go/tools/bazel:go_default_library",
        ],
        data = test_data + [
            gazelle_binary,
            ":generation_test_manifest.yaml",
        ],
```

Reply:
Yep, this seems much better.

> This can be a
>           https://docs.bazel.build/versions/main/be/general.html#filegroup

Looks like that doc has changed locations to https://bazel.build/reference/be/general#filegroup

Reply:
Hm another issue is that `$(location %s)` gives a workspace relative path, when I need an absolute path (I get `gazelle_local_/gazelle_local: no such file or directory`).

How do I use that value cleanly?

Reply:
Ah, looks like I can use https://github.com/bazelbuild/rules_go/blob/v0.30.0/go/tools/bazel/runfiles.go#L43